design/sketch.webidl

```diff
@@ -0,0 +1,532 @@
+typedef u32 WebGPUPrimitiveTopologyEnum;
+interface WebGPUPrimitiveTopology {
+    const u32 POINT_LIST = 0;
+    const u32 LINE_LIST = 1;
+    const u32 LINE_STRIP = 2;
+    const u32 TRIANGLE_LIST = 3;
+    const u32 TRIANGLE_STRIP = 4;
+};
+
+dictionary WebGPURenderPipelineDescriptor : WebGPUPipelineDescriptorBase{
+    WebGPUPrimitiveTopologyEnum primitiveTopology;
+    sequence<WebGPUBlendState> blendState;
+    WebGPUDepthStencilState depthStencilState;
+    WebGPUInputState inputState;
+    WebGPUAttachmentState attachmentState;
+    // TODO other properties
+};
+
+interface WebGPURenderPipeline {
+};
+// ****************************************************************************
+// COMMAND RECORDING (Command buffer and all relevant structures)
+// ****************************************************************************
+
+typedef u32 WebGPULoadOpEnum;
+interface WebGPULoadOp {
+    const u32 CLEAR = 0;
+    const u32 LOAD = 1;
+};
+
+typedef u32 WebGPUStoreOpEnum;
+interface WebGPUStoreOp {
+    const u32 STORE = 0;
+};
+
+dictionary WebGPURenderPassAttachmentDescriptor {
+    WebGPUTextureView attachment;
+    WebGPULoadOp loadOp;
+    WebGPUStoreOp storeOp;
+}
+
+dictionary WebGPURenderPassDescriptor {
+    sequence<WebGPURenderPassAttachmentDescriptor> colorAttachments;
+    WebGPURenderPassAttachmentDescriptor depthStencilAttachmaent;
+};
+
+interface WebGPUCommandBuffer {
+};
+
+interface WebGPUCommandEncoder {
+    WebGPUCommandBuffer finishEncoding();
+
+    // Commands allowed outside of "passes"
+    void copyBufferToBuffer(WebGPUBuffer src,
+                            u32 srcOffset,
+                            WebGPUBuffer dst,
+                            u32 dstOffset,
+                            u32 size);
+    // TODO figure out all the arguments required for these
+    void copyBufferToTexture();
+    void copyTextureToBuffer();
+    void copyTextureToTexture();
+    void blit();
+
+    // Allowed in both compute and render passes
+    void setPushConstants(WebGPUShaderStageFlags stage,
+                          u32 offset,
+                          u32 count,
+                          ArrayBuffer data);
+    void setBindGroup(u32 index, WebGPUBindGroup bindGroup);
+    void setPipeline((WebGPUComputePipeline or WebGPURenderPipeline) pipeline);
+
+    // Compute pass commands
+    void beginComputePass();
+    void endComputePass();
+
+    void dispatch(u32 x, u32 y, u32 z);
+
+    // Render pass commands
+    void beginRenderPass(WebGPURenderPassDescriptor descriptor);
+    void endRenderPass();
+
+    void setBlendColor(float r, float g, float b, float a);
+    void setIndexBuffer(WebGPUBuffer buffer, u32 offset);
+    void setVertexBuffers(u32 startSlot, sequence<WebGPUBuffer> buffers, sequence<u32> offsets);
+
+    void draw(u32 vertexCount, u32 instanceCount, u32 firstVertex, u32 firstInstance);
+    void drawIndexed(u32 indexCount, u32 instanceCount, u32 firstIndex, u32 firstInstance);
```

Comment:
Minor thing, but let's not forget `baseIndex` (or `fistVertex`, as in Vulkan).

Reply:
nit: The used term in D3D12 and Vulkan is `baseVertex`

Reply:
Done